plugin/quic/setup.go

```diff
@@ -0,0 +1,151 @@
+package quic
+
+import (
+	"crypto/tls"
+	"fmt"
+
+	"github.com/coredns/caddy"
+	"github.com/coredns/coredns/core/dnsserver"
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/plugin/pkg/parse"
+	pkgtls "github.com/coredns/coredns/plugin/pkg/tls"
+)
+
+func init() { plugin.Register("quic", setup) }
+
+func setup(c *caddy.Controller) error {
+	q, err := ParseQUIC(c)
+	if err != nil {
+		return plugin.Error("quic", err)
+	}
+
+	if q.len() > max {
+		return plugin.Error("quic", fmt.Errorf("more than %d TOs configured: %d", max, q.len()))
+	}
+
+	dnsserver.GetConfig(c).AddPlugin(func(next plugin.Handler) plugin.Handler {
+		q.Next = next // Set the Next field, so the plugin chaining works.
+		return q
+	})
+
+	return nil
+}
+
+func ParseQUIC(c *caddy.Controller) (*QUIC, error) {
+	var (
+		q   *QUIC
+		err error
+		i   int
+	)
+	for c.Next() {
+		if i > 0 {
+			return nil, plugin.ErrOnce
+		}
+		i++
+		q, err = parseStanza(c)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return q, nil
+}
+
+func parseStanza(c *caddy.Controller) (*QUIC, error) {
+	q := NewQUIC()
+
+	if !c.Args(&q.from) {
+		return q, c.ArgErr()
+	}
+	normalized := plugin.Host(q.from).NormalizeExact()
+	if len(normalized) == 0 {
+		return q, fmt.Errorf("unable to normalize '%s'", q.from)
+	}
+	q.from = normalized[0] // only the first is used.
+
+	to := c.RemainingArgs()
+	if len(to) == 0 {
+		return q, c.ArgErr()
+	}
+
+	toHosts, err := parse.HostPortOrFile(to...)
+	if err != nil {
+		return q, err
+	}
+
+	for c.NextBlock() {
+		if err := parseBlock(c, q); err != nil {
+			return q, err
+		}
+	}
+
+	if q.tlsConfig == nil {
+		q.tlsConfig = new(tls.Config)
+	}
+
+	if q.tlsServerName != "" {
+		q.tlsConfig.ServerName = q.tlsServerName
+	} else {
+		q.tlsConfig.InsecureSkipVerify = true
```

Comment:
I'll have to find a way to circumvent this. `quic-go` doesn't work without `tls.Config` and I couldn't set the `tls_servername` using `plugin/tls` test certificates.

Reply:
@jaehnri the plugin/dns test certs common name is empty, therefore the hostname cannot match. My suggestion would be to drop the tls_servername option, because it should never be needed (the tls package checks if the common name, subject alternative names or ip addresses contains the hosts name) and instead introduce the option tls_skip_verify. For test cases u can set the common name manually via the tlsConfig